core/src/main/java/org/elasticsearch/rest/action/bulk/RestBulkAction.java

```diff
@@ -97,49 +98,35 @@ public RestResponse buildResponse(BulkResponse response, XContentBuilder builder
-                        if (itemResponse.getResponse() instanceof DeleteResponse) {
-                            DeleteResponse deleteResponse = itemResponse.getResponse();
-                            if (deleteResponse.isFound()) {
-                                builder.field(Fields.STATUS, shardInfo.status().getStatus());
-                            } else {
-                                builder.field(Fields.STATUS, RestStatus.NOT_FOUND.getStatus());
+                        final DocWriteResponse docResponse = itemResponse.getResponse();
+                        docResponse.toXContent(builder, request);
+                        RestStatus status = docResponse.getShardInfo().status();
+                        if (docResponse instanceof DeleteResponse) {
+                            DeleteResponse deleteResponse = (DeleteResponse) docResponse;
+                            if (deleteResponse.isFound() == false) {
+                                status = RestStatus.NOT_FOUND;
                             }
-                            builder.field(Fields.FOUND, deleteResponse.isFound());
-                        } else if (itemResponse.getResponse() instanceof IndexResponse) {
-                            IndexResponse indexResponse = itemResponse.getResponse();
+                        } else if (docResponse instanceof IndexResponse) {
+                            IndexResponse indexResponse = (IndexResponse) docResponse;
                             if (indexResponse.isCreated()) {
-                                builder.field(Fields.STATUS, RestStatus.CREATED.getStatus());
-                            } else {
-                                builder.field(Fields.STATUS, shardInfo.status().getStatus());
+                                status = RestStatus.CREATED;
                             }
-                        } else if (itemResponse.getResponse() instanceof UpdateResponse) {
-                            UpdateResponse updateResponse = itemResponse.getResponse();
+                        } else if (docResponse instanceof UpdateResponse) {
+                            UpdateResponse updateResponse = (UpdateResponse) docResponse;
                             if (updateResponse.isCreated()) {
-                                builder.field(Fields.STATUS, RestStatus.CREATED.getStatus());
-                            } else {
-                                builder.field(Fields.STATUS, shardInfo.status().getStatus());
-                            }
-                            if (updateResponse.getGetResult() != null) {
-                                builder.startObject(Fields.GET);
-                                updateResponse.getGetResult().toXContentEmbedded(builder, request);
-                                builder.endObject();
+                                status = RestStatus.CREATED;
                             }
                         }
+                        builder.field(Fields.STATUS, status.getStatus());
```

Comment:
can we add an abstract method getRestStatus() to DocWriteResponse (to be overwritten by its subclasses) so that we don't need all these instanceof...